packages/next/src/server/lib/router-utils/setup-dev.ts

```diff
@@ -1236,11 +1232,23 @@ async function startWatcher(opts: SetupOpts) {
 
   opts.fsChecker.ensureCallback(async function ensure(item) {
     if (item.type === 'appFile' || item.type === 'pageFile') {
+      // If the type is an app file, then we should mark it as an app page.
+      const isApp = item.type === 'appFile'
+
+      // Try to get the page definition for the item.
+      const definition = isApp
+        ? opts.fsChecker.appRoutes.get(item.itemPath)
+        : null
+
       await hotReloader.ensurePage({
+        page: definition?.page ?? item.itemPath,
         clientOnly: false,
-        page: item.itemPath,
-        isApp: item.type === 'appFile',
+        definition,
```

Comment:
This is the magic sauce, from here we're now able to throw the definition into the ensure function allowing us to get critical information (like the `appPaths`) that is needed for compiling App Pages.